src/protocols/secure_channel/SessionResumptionStorage.cpp

```diff
@@ -0,0 +1,151 @@
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the CHIP CASE Session object that provides
+ *      APIs for constructing a secure session using a certificate from the device's
+ *      operational credentials.
+ */
+
+#include <protocols/secure_channel/SessionResumptionStorage.h>
+
+#include <lib/support/Base64.h>
+#include <lib/support/SafeInt.h>
+
+namespace chip {
+
+CHIP_ERROR SessionResumptionStorage::FindByScopedNodeId(const ScopedNodeId & node, ResumptionIdStorage & resumptionId,
+                                                        Crypto::P256ECDHDerivedSecret & sharedSecret, CATValues & peerCATs)
+{
+    ReturnErrorOnFailure(LoadState(node, resumptionId, sharedSecret, peerCATs));
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SessionResumptionStorage::FindByResumptionId(ConstResumptionIdView resumptionId, ScopedNodeId & node,
+                                                        Crypto::P256ECDHDerivedSecret & sharedSecret, CATValues & peerCATs)
+{
+    ReturnErrorOnFailure(FindNodeByResumptionId(resumptionId, node));
+    ResumptionIdStorage tmpResumptionId;
+    ReturnErrorOnFailure(FindByScopedNodeId(node, tmpResumptionId, sharedSecret, peerCATs));
+    VerifyOrReturnError(std::equal(tmpResumptionId.begin(), tmpResumptionId.end(), resumptionId.begin(), resumptionId.end()),
+                        CHIP_ERROR_KEY_NOT_FOUND);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SessionResumptionStorage::FindNodeByResumptionId(ConstResumptionIdView resumptionId, ScopedNodeId & node)
+{
+    ReturnErrorOnFailure(LoadLink(resumptionId, node));
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SessionResumptionStorage::Save(const ScopedNodeId & node, ConstResumptionIdView resumptionId,
+                                          const Crypto::P256ECDHDerivedSecret & sharedSecret, const CATValues & peerCATs)
+{
+    SessionIndex index;
+    ReturnErrorOnFailure(LoadIndex(index));
+
+    if (index.mSize == CHIP_CONFIG_CASE_SESSION_RESUME_CACHE_SIZE)
+    {
+        // TODO: implement LRU for resumption
+        ReturnErrorOnFailure(Delete(index.mNodes[0]));
+        ReturnErrorOnFailure(LoadIndex(index));
+    }
+
+    ReturnErrorOnFailure(SaveState(node, resumptionId, sharedSecret, peerCATs));
+    ReturnErrorOnFailure(SaveLink(resumptionId, node));
+
+    index.mNodes[index.mSize++] = node;
+    ReturnErrorOnFailure(SaveIndex(index));
```

Comment:
I had approved this PR because the code now has a capability to limit the size of this feature in storage.

However, this is really complex for what it does.  We have three tables that must remain in sync, and if they don't, we will permanently leak storage.  And while the code is attempting to give O(1) performance, it actually doesn't because we still need to store the entire index for every write.

We could very well do this with a single set of entries of this key=value shape:

`g/r/<mask<resumptionId>>=<fabricIndex, nodeId, secret, CATs, resumptionId>`

You can make the mask as small or large as you like to suit your platform.  You get O(1) lookups for sigma2_resume messages coming in the door, O(N) lookups when establishing outbound sessions, and O(1) for all writes.  And importantly, this scheme is less likely to leak storage.

I know that O(N) lookup for outbound session establishment doesn't sound great, but it's a trade off for simplicity and robustness.  Also, read may be much faster than write on some platforms.  And because this penalty is for outbound session creation, there's not really any security concern.